pandas/core/indexes/base.py

```diff
@@ -2912,7 +2913,22 @@ def union(self, other, sort=None):
         self._assert_can_do_setop(other)
         other, result_name = self._convert_can_do_setop(other)
 
-        if self.dtype != other.dtype:
+        if isinstance(self.dtype, CategoricalDtype) and isinstance(
+            other.dtype, CategoricalDtype
+        ):
+            both_categories = self.categories  # type: ignore[attr-defined]
+            if len(self.categories) != len(other.categories) or any(  # type: ignore[attr-defined]
+                self.categories != other.categories  # type: ignore[attr-defined]
+            ):
+                # Unite both categories
+                both_categories = np.union1d(self.categories, other.categories)  # type: ignore[attr-defined]
```

Comment:
You should be able to use `union_categoricals` in `pandas/core/dtypes/concat.py`

Reply:
I've just made the change in the last commit.